src/Compilers/CSharp/Portable/Syntax/SyntaxExtensions.cs

```diff
@@ -387,5 +387,59 @@ internal static bool IsOutVarDeclaration(this DeclarationExpressionSyntax p)
         {
             return p.Designation.Kind() == SyntaxKind.SingleVariableDesignation && p.IsOutDeclaration();
         }
+
+        /// <summary>
+        /// Visits all the ArrayRankSpecifiers of a typeSyntax, invoking an action on each one in turn
+        /// </summary>
+        /// <param name="type"></param>
+        /// <param name="action"></param>
+        /// <param name="argument">The argument that is passed to the action whenever it is invoked</param>
+        internal static void VisitRankSpecifiers<TArg>(this TypeSyntax type, Action<ArrayRankSpecifierSyntax, TArg> action, TArg argument)
+        {
+            switch (type)
+            {
+                case ArrayTypeSyntax arrayTypeSyntax:
+                    arrayTypeSyntax.ElementType.VisitRankSpecifiers(action, argument);
+                    foreach (var rankSpecifier in arrayTypeSyntax.RankSpecifiers)
+                    {
+                        action(rankSpecifier, argument);
+                    }
+                    break;
+                case NullableTypeSyntax nullableTypeSyntax:
+                    nullableTypeSyntax.ElementType.VisitRankSpecifiers(action, argument);
+                    break;
+                case PointerTypeSyntax pointerTypeSyntax:
+                    pointerTypeSyntax.ElementType.VisitRankSpecifiers(action, argument);
+                    break;
+                case TupleTypeSyntax tupleTypeSyntax:
+                    foreach (var element in tupleTypeSyntax.Elements)
+                    {
+                        element.Type.VisitRankSpecifiers(action, argument);
+                    }
+                    break;
+                case RefTypeSyntax refTypeSyntax:
+                    refTypeSyntax.Type.VisitRankSpecifiers(action, argument);
+                    break;
+                case GenericNameSyntax genericNameSyntax:
+                    foreach (var typeArgument in genericNameSyntax.TypeArgumentList.Arguments)
+                    {
+                        typeArgument.VisitRankSpecifiers(action, argument);
+                    }
+                    break;
+                case QualifiedNameSyntax qualifiedNameSyntax:
+                    qualifiedNameSyntax.Right.VisitRankSpecifiers(action, argument);
```

Comment:
>qualifiedNameSyntax.Right [](start = 20, length = 25)

What about the ```Left```?

Reply:
I forgot that could also have an array in it...